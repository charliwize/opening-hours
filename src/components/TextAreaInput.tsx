import { ChangeEvent } from 'react';
import styled from 'styled-components';

interface Props {
    value: string;
    handleEnterData: (e: ChangeEvent<HTMLTextAreaElement>) => void;
    getSchedules: () => void;
    error?: string;
}

const StyledForm = styled.form`
    display: flex;
    flex-direction: column;
    padding-bottom: 1.5rem;
    width: 100%;
    textarea {
        padding: 1.5rem;
        border: 1.5px solid #202125;
    }
    h1 {
        border: 0;
        clip: rect(1px 1px 1px 1px);
        clip; rect(1px, 1px, 1px, 1px);
        height: 1px;
        margin: -1px;
        overflow: hidden;
        padding: 0;
        position: absolute;
        width: 1px; 
    }
    label {
        color: #A1A2A4;
        font-size: 0.75rem;
        margin-bottom: 0.5rem;
    }
`;

const ErrorText = styled.div`
    width: fit-content;
    color: red;
    font-weight: bold;
`;

const StyledButton = styled.button`
    margin-top: 1rem;
    display: flex;
    width: fit-content;
    min-width: 180px;
    background: #202125;
    justify-content: center;
    color: #fff;
    border: 0;
    padding: 0.5rem;
    font-weight: 600;
    font-size: 0.75rem;
    text-transform: uppercase;
    @media (max-width: 320px) {
        width: 100%;
    }
    :hover {
        color: #A1A2A4;
    }
    :disabled {
        background: #EEEEEE;
        color: #A1A2A4;
    }
`;

const ButtonContainer = styled.div`
    display: flex;
    flex-direction: column;
`;

const TextAreaInput = ({ value, error, handleEnterData, getSchedules }: Props) => {
    return (
        <StyledForm onSubmit={(e) => e.preventDefault()}>
            <h1>Insert JSON Here</h1>
            <label htmlFor="json-input-textarea">
                paste JSON below, click 'GET SCHEDULE' to view updated opening hours
            </label>
            <textarea
                name="text-area"
                id="json-input-textarea"
                value={value}
                onChange={handleEnterData}
                rows={20}
            />
            <ButtonContainer>
                <StyledButton
                    type="submit"
                    onClick={getSchedules}
                    disabled={!!error || value.length === 0}
                >
                    <span>Get Schedule</span>
                </StyledButton>
            </ButtonContainer>
            {error && <ErrorText>{error}</ErrorText>}
        </StyledForm>
    );
};

export default TextAreaInput;
