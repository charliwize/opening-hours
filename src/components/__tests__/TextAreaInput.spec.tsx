import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import TextAreaInput from '../TextAreaInput';

const defaultProps = {
    value: 'dummy_json',
    handleEnterData: jest.fn(),
    getSchedules: jest.fn(),
    error: 'string',
};

describe('TextAreaInput', () => {
    test('it renders', () => {
        const { container } = render(
            <TextAreaInput
                handleEnterData={defaultProps.handleEnterData}
                getSchedules={defaultProps.getSchedules}
                value={defaultProps.value}
            />
        );
        expect(
            screen.getByLabelText(
                /paste JSON below, click 'GET SCHEDULE' to view updated opening hours/
            )
        ).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    test('it enables button on textarea changed', () => {
        render(
            <TextAreaInput
                handleEnterData={defaultProps.handleEnterData}
                getSchedules={defaultProps.getSchedules}
                value=""
            />
        );
        const input = screen.getByLabelText(
            /paste JSON below, click 'GET SCHEDULE' to view updated opening hours/
        );
        expect(screen.queryAllByRole('button')[0]).toBeDisabled();
        fireEvent.change(input, { target: { value: 123 } });
        waitFor(() => {
            expect(screen.getByRole('button')).toBeEnabled();
        });
    });

    test('it gets schedules on button clicked', async () => {
        render(
            <TextAreaInput
                handleEnterData={defaultProps.handleEnterData}
                getSchedules={defaultProps.getSchedules}
                value="123"
            />
        );
        const button = screen.queryAllByRole('button')[0];
        expect(button).toBeEnabled();
        fireEvent.click(button);
        expect(defaultProps.getSchedules).toHaveBeenCalledTimes(1);
    });
});
