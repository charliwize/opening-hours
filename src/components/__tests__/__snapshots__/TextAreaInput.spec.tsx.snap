// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaInput it renders 1`] = `
<div>
  <form
    class="sc-bdvvtL fGZnOR"
  >
    <h1>
      Insert JSON Here
    </h1>
    <label
      for="json-input-textarea"
    >
      paste JSON below, click 'GET SCHEDULE' to view updated opening hours
    </label>
    <textarea
      id="json-input-textarea"
      name="text-area"
      rows="20"
    >
      dummy_json
    </textarea>
    <div
      class="sc-hKwDye fnnqVZ"
    >
      <button
        class="sc-dkPtRN drFYsO"
        type="submit"
      >
        <span>
          Get Schedule
        </span>
      </button>
    </div>
  </form>
</div>
`;
